petastorm/reader.py

```diff
@@ -454,18 +454,42 @@ def __init__(self, pyarrow_filesystem, dataset_path, schema_fields=None,
         # 4. Create a rowgroup ventilator object
         normalized_shuffle_row_drop_partitions = \
             self._normalize_shuffle_options(shuffle_row_drop_partitions, self.dataset)
-        ventilator = self._create_ventilator(filtered_row_group_indexes, shuffle_row_groups,
-                                             normalized_shuffle_row_drop_partitions, num_epochs, worker_predicate,
-                                             self._workers_pool.workers_count + _VENTILATE_EXTRA_ROWGROUPS)
+        self.ventilator = self._create_ventilator(filtered_row_group_indexes, shuffle_row_groups,
+                                                  normalized_shuffle_row_drop_partitions, num_epochs, worker_predicate,
+                                                  self._workers_pool.workers_count + _VENTILATE_EXTRA_ROWGROUPS)
 
         # 5. Start workers pool
         self._workers_pool.start(worker_class, (pyarrow_filesystem, dataset_path, storage_schema, self.ngram,
                                                 row_groups, cache, transform_spec),
-                                 ventilator=ventilator)
+                                 ventilator=self.ventilator)
         logger.debug('Workers pool started')
 
         self.last_row_consumed = False
 
+    def reset(self):
+        """Resets ``Reader`` state and allows to fetch more samples once the ``Reader`` finished reading all epochs,
+        as specified by the ``num_epochs`` parameter.
+
+        Once all samples were read from a reader, an attempt to fetch new sample (e.g. ``next(reader)`` would raise
+        ``StopIterationError``. You can reset the reader to the original state and restart reading samples
+        calling ``reset()``.
+
+        We do not support calling ``reset()`` until all samples were consumed. ``NotImplementedError``
+        will be raised if a user attempt to do so.
+
+        :return: None
+        """
+        # TODO(yevgeni): could there be a race here?
+        if not self.last_row_consumed:
+            # Don't allow reseting in the middle of epoch iterations since it is not very well defined how
```

Comment:
I feel like if somebody wanted to reset in the middle of an iteration it would be fine to drop the in flight samples (as that is the behavior I would expect). Though I can't think of a reason somebody would need that so I think its fine to not implement that for now (especially since there are likely race conditions with that one)

Reply:
Agree. My train of thought was that it would be hard to guarantee anything (either we drop, or we don't, or drop some?). A clean way is probably to drop existing workers and existing output queue and replace them with new ones (that would make sure we kill all inflights). Wanted to avoid this kind of undefined behavior.